New payment slip for v8

nbessi opened the pull request.

nbessi committed:
Move payment slip

nbessi committed:
Port bank to new api

nbessi committed:
Port company to new API

nbessi committed:
Extracting payment slip related knowledge

into a new model dedicated to payment slip

nbessi committed:
Moving media into static

nbessi committed:
Changing company defaults for payment_slip

nbessi committed:
Moving font file into static

nbessi committed:
Adding QWeb report skeletton

nbessi committed:
Payment slip is know draw using PIL

nbessi committed:
Fix New API syntax error when proting and remove dead files

nbessi committed:
Fix invalid variable name

nbessi committed:
Import payment slip model

nbessi committed:
Add first version of new payment slip report

nbessi committed:
Adjusting Slip element position

nbessi committed:
Adjust amount placement

nbessi committed:
Implementing new report parser and adding a4 support

nbessi committed:
Fix resize and pdf resolution

nbessi committed:
Removing comment test are going to be ported

nbessi committed:
Implementing bvr references fields and adding doc strings

nbessi committed:
Adding readonly property on slip fields and sql constraints

nbessi committed:
Adding docstring

nbessi committed:
Porting v1 import wizard

guewen commented on l10n_ch_payment_slip/payment_slip.py:
```diff
@@ -0,0 +1,635 @@
+        draw.text((x, y), ref, font=font, fill=self._fill_color)
+
+    @api.model
+    def _draw_amont(self, draw, font, amount, initial_position, company):
```
s/_draw_amont/_draw_amount/

guewen commented on l10n_ch_payment_slip/report/payment_slip_from_invoice.py:
```diff
@@ -0,0 +1,81 @@
+            raise
+        finally:
+            buff.close()
+            [x.close() for x in streams]
```
This is an abuse of list comprehensions. They should only be used to create a list.
Better: 

```
for stream in streams:
    stream.close()
```

guewen commented on l10n_ch_payment_slip/wizard/bvr_import.py:
```diff
@@ -0,0 +1,251 @@
+        values = {'name': '/',
+                  'date': record['date'],
+                  'amount': record['amount'],
+                  'ref': reference,
```
In fact the name is used for the matching so the reference should goes there I think

nbessi committed:
Adding basic test on payment slip

guewen commented on l10n_ch_payment_slip/invoice.py:
```diff
@@ -0,0 +1,155 @@
+    _name = 'account.tax.code'
+    _inherit = "account.tax.code"
+
+    code = fields.Char('Case Code', size=512),
```
Comma at end of line. I wanted to correct it on your branch but I haven't the push rights.
Also, the string 'Case Code' can be removed because it is already defined in the 'account' module.

guewen commented on l10n_ch_payment_slip/payment_slip.py:
```diff
@@ -0,0 +1,635 @@
+        line.append('+')
+        line.append(' ')
+        bank = self.move_line_id.invoice.partner_bank_id.get_account_number()
+        account_components = bank.split('-')
```
```
File "/home/gbaconnier/code/dev/parts/l10n-switzerland/l10n_ch_payment_slip/payment_slip.py", line 218, in _compute_scan_line_list
    account_components[1].rjust(6, '0'),
ValueError: "list index out of range" while evaluating
```

if `get_account_number()` returns an account in the wrong format (no `-` in mine for instance).

nbessi replied:
Fixed

nbessi committed:
Typo _draw_amont -> _draw_amount

nbessi committed:
Scan line must not be stored

nbessi committed:
Fix v11 import reference goes in name

and invoice goes in ref

nbessi committed:
Fix base bank constrains and add unit test

nbessi committed:
Improve code deffensivness

nbessi committed:
Add payment slip unit test

nbessi committed:
Add fast suite

nbessi committed:
Add account as direct dependence of payment slip

nbessi committed:
Fix dead code

nbessi committed:
RM dead code

nbessi committed:
Fix on change partner_id

nbessi committed:
Add print of bank

nbessi committed:
Fix ref field updated on validation update

nbessi committed:
Slip button in `group_system` (Settings)

nbessi committed:
Better spacing

nbessi committed:
PEP8

nbessi committed:
PEP8 PYlint

nbessi committed:
Better manifest spacing

nbessi committed:
Fix action order

nbessi committed:
Imp loop styling

nbessi commented on l10n_ch_payment_slip/report/payment_slip_from_invoice.py:
```diff
@@ -0,0 +1,84 @@
+            raise
+        finally:
+            buff.close()
+            for stream in streams:
```
Agreed and fixed

nbessi committed:
Fix import order

nbessi committed:
Add options to merge PDF's

It can be done on disk or in memory

nbessi committed:
Add merge option in views

nbessi committed:
PEP8

nbessi committed:
Remove deprecated tests

nbessi committed:
Add common in import statement

nbessi committed:
Porting l10n_ch_credit_control_payment_slip_report

nbessi committed:
PEP8

nbessi committed:
Fix amount computation

nbessi committed:
Better model name and descritpions

nbessi committed:
Add v11 import tests

nbessi committed:
Fix comments and authors

nbessi committed:
Add credit control payment slip integration tests

nbessi committed:
Add payment slip security

nbessi committed:
Improve manifest

nbessi committed:
Add credit control integration security

nbessi committed:
Fix compute full name must be empty depending of invoice state

nbessi committed:
Fix error string

nbessi committed:
Fix error message

nbessi committed:
Add translation of payment slip

nbessi committed:
Fix translation files

nbessi committed:
Fix tests on travis in case of slow process

nbessi committed:
Greenify tests

guewen commented:
The module `base_transaction_id` should be added as dependency because it is required in order to reconcile the `transaction_ref` with the new bank statement.

nbessi committed:
Fixing payment slip dependencies

Removing report_webkit adding base_transaction_id

guewen commented:
Thanks :+1:

guewen commented on l10n_ch_base_bank/bank.py:
```diff
@@ -227,7 +221,7 @@ def _check_postal_num(self):
                 )
 
-    @api.constrains('acc_number', 'bank')
+    @api.constrains('acc_number', 'bank.ccp')
```
```
openerp.models: @constrains('acc_number', 'bank.ccp') parameters must be field names
```

AFAIK `@api.constrains()` does not support dotted notation, only fields of the current model.

nbessi replied:
@guewen right you are, we have to add a constrains on res.bank

Commit:
Constrains must includes only fields of the current model

Anyway, the check already exist on the res.bank model.  And we want to check
the ccp duplication when it is changed or the bank is changed, so
('acc_number', 'bank') is correct.

nbessi commented:
needs a rebase

lepistone commented on .travis.yml:
```diff
@@ -21,8 +21,7 @@ install:
 
 after_success:
-  coveralls
+  coveralls
```
no newline at end of file

lepistone commented on l10n_ch_base_bank/tests/__init__.py:
```diff
@@ -0,0 +1,24 @@
+#
+##############################################################################
+from . import test_bank
+fast_suite = [
```
fast_suite, full suite and checks are ignored in v8

nbessi replied:
fixed

lepistone commented on l10n_ch_base_bank/tests/test_bank.py:
```diff
@@ -0,0 +1,150 @@
+            }
+        )
+
+    def test_faulty_ccp_at_bank(self):
```
not blocking over that, but a way to test that in a couple lines is to create your object with `new()` with only the field that is checked and then to call the check method directly.

lepistone commented on l10n_ch_credit_control_payment_slip_report/payment_slip.py:
```diff
@@ -0,0 +1,43 @@
+from openerp import models
+
+
+class payment_slip(models.Model):
```
PaymentSlip

nbessi replied:
Don't feed the troll ;P

lepistone replied:
OK :smile_cat:

lepistone commented on l10n_ch_credit_control_payment_slip_report/tests/__init__.py:
```diff
@@ -19,14 +18,8 @@
-            'res.company', 'partner_id',
-            'Companies that refers to partner'),
-    }
+fast_suite = [
```
as above, this is ignored

nbessi replied:
fixed

lepistone commented on l10n_ch_credit_control_payment_slip_report/tests/test_payment_slip_report.py:
```diff
@@ -0,0 +1,149 @@
+        attempt = 0
+        while not invoice.move_id:
+            invoice.refresh()
+            time.sleep(0.1)
```
can you document in a comment why this hack is necessary?

nbessi replied:
Workflows are asynchrone

nbessi replied:
done

lepistone commented on l10n_ch_credit_control_payment_slip_report/tests/test_payment_slip_report.py:
```diff
@@ -0,0 +1,149 @@
+
+class TestPaymentSlipReport(test_common.TransactionCase):
+
+    def make_bank(self):
```
not blocking, but since each test is in a different transaction, you can also do your setup in `setUp()` and store it as class attributes, without the need for helper methods. (just a matter of taste, you might prefer helper methods)

lepistone replied:
sorry @nbessi . I meant instance attribute, not class attribute. I meant

``` python
class TestThis():
    def setUp():
        self.order = Order.create({...})
```

lepistone commented on l10n_ch_payment_slip/__openerp__.py:
```diff
@@ -26,52 +26,53 @@
-but will remain in the data model via new banking addons dependence.
 
 """,
  'version': '1.2',
```
not blocking. The module changed a bit more than just the porting to v8. I'd bump this version.

nbessi replied:
Fixed

nbessi committed:
Remove deprecated fast_suite

nbessi committed:
Removing deprecated fast_suite

nbessi committed:
Update addons version

nbessi committed:
Add comment on test workflow trick

lepistone commented:
A few minor remarks, but :+1: thanks!

guewen commented on l10n_ch_payment_slip/payment_slip.py:
```diff
@@ -0,0 +1,667 @@
+
+    amount_total = fields.Float('Total amount of BVR/ESR',
+                                compute='compute_amount',
+                                store=True)
```
If `amount_total` is stored, the payment slip with dunning fees won't include the fees: it won't recompute the `total_amount` so the fees won't be included. @nbessi, what is the reason to store the total?

nbessi commented on l10n_ch_credit_control_payment_slip_report/credit_control_line.py:
```diff
@@ -18,23 +18,26 @@
 
         :param move: move line report
-        :param rtype: report type string just in case
+        :param rtype: report type string just cin case
```
typo cin/in

nbessi commented:
@Guewen mostly for perfomance and to enable future search, group by etc. The  `propagate_fees` should ensure that total must be written there is mybe a bug with it or a modification in v8 API.

guewen commented:
> @Guewen mostly for perfomance and to enable future search, group by etc. The propagate_fees should ensure that total must be written there is mybe a bug with it or a modification in v8 API

I don't see use cases for searches or group by on payment slips. `propagate_fees` is a `@api.constrains`, it is not meant for updating values. 

It does not work because when `propagate_fees` the credit control line is in draft and the `
_compute_amount_hook` has the domain `[('move_line_id', '=', self.move_line_id.id), ('state', 'in', ('to_be_sent', 'sent'))]`, so the amount is not updated and won't be updated again.

I remove the store on the field. Performance is not really a concern as the function is only `self.move_line_id.debit`.

nbessi commented:
@guewen Ok let keep it simple. If needs appear to store the amount we will adapt the code

Commit:
Wrong exception raised

Commit:
The amount total on the payment slip is no longer stored

This change allows to have a dynamic amount, for instance when
we want to add the dunning fees.

Commit:
Complete the dependencies and use the field in the scan line

Commit:
Update translations for l10n_ch_credit_control_payment_slip_report

Commit:
Update translations for l10n_ch_payment_slip

Commit:
Update translations in l10n_ch_base_bank

Commit:
Fix the dunning fees test

* Pass the id of the credit line to add in the total in the context
* Extends the depends of the computed field otherwise the first value computed
  without the fees is kept in cache

Commit:
Remove view that breaks the onchange because it replaces the whole field

Remove it entirely because 'select="1"' has no effect on views now that we use
search views.

Commit:
Do not use 'replace' position which can lose attributes

Only change the domain attribute

Commit:
Remove deprecated type of view

Commit:
onchange_partner_bank breaks the base onchange

Because it is declared as api.onchange whereas the base method is an
api.multi with the old-api style 'on_change' in XML. If we declare
onchange_partner_bank as api.onchange, it overrides the base one and
that's not what we want.

Commit:
Merge commit 'refs/pull/75/head' of github.com:OCA/l10n-switzerland into new_payment_slip

Conflicts:
	l10n_ch_base_bank/invoice.py

guewen commented:
Rebased on top of 8.0.
And merged these fixes #75

Commit:
Remove formatting in transaction_ref of move lines

When we import V11 files, the bvr numbers are not formatted,
we find them using the unformatted number

Commit:
Add a space between bvr numbers on the invoice for readability

tafaRU committed:
Fix issue #80

nbessi committed:
Merge pull request #4 from tafaRU/8.0-l10n_ch_payment_slip-fix-issue-80

[Fix] issue #80

tafaRU commented:
@nbessi could you please rebase on 8.0 to include last fixes released on **_l10n_ch_dta**_?

Thank you in advance.

nbessi commented:
@tarafu I will prefer to have a merge and treat new problem as seprated issues.

tafaRU committed:
Fix https://github.com/OCA/l10n-switzerland/issues/82

nbessi committed:
Merge pull request #5 from tafaRU/8.0-l10n_ch_payment_slip-fix-issue-82

Fix issue #82

eLBati commented:
Hello @nbessi , I'm testing `l10n_ch_payment_slip` with latest odoo 8 code and trying to print a payment slip from invoice, I always get an empty file.
I noticed `_draw_payment_slip` does not even get executed. Does this happen to you?
Thanks

nbessi commented:
@eLBati no I do not encounterd this problem maybe you have a PIL dependency missing or an other report engine that may conflict.

Commit:
Only the first payment slip is printed on credit lines
